packages/client/src/runtime/core/engines/common/resolveEnginePath.ts

```diff
@@ -0,0 +1,125 @@
+  // if the user provided a custom path, or if engine previously found
+  const prismaPath = {
+    binary: process.env.PRISMA_QUERY_ENGINE_BINARY ?? config.prismaPath,
+    library: process.env.PRISMA_QUERY_ENGINE_LIBRARY ?? config.prismaPath,
+  }[engineType]
+
+  if (prismaPath !== undefined) return prismaPath
+
+  // otherwise we will search to find the nearest query engine file
+  const { enginePath, searchedLocations } = await findEnginePath(engineType, config)
+
+  debug('enginePath', enginePath)
+
+  // if we find it, we apply +x chmod to the binary, cache, and return
+  if (enginePath !== undefined && engineType === 'binary') plusX(enginePath)
+  if (enginePath !== undefined) return (config.prismaPath = enginePath)
+
+  // if we don't find it, then we will throw helpful error messages
+  const binaryTarget = await getPlatform()
+  const generatorBinaryTargets = config.generator?.binaryTargets ?? []
+  const hasNativeBinaryTarget = generatorBinaryTargets.some((bt) => bt.native === true)
+  const hasMissingBinaryTarget = !generatorBinaryTargets.some((bt) => bt.value === binaryTarget)
+  const clientHasBeenBundled = __filename.match(/library|binary\.m?js/) === null // runtime name
+
+  const errorInput: EngineNotFoundErrorInput = {
+    searchedLocations,
+    generatorBinaryTargets,
+    generator: config.generator!,
+    runtimeBinaryTarget: binaryTarget,
+    queryEngineName: getQueryEngineName(engineType, binaryTarget),
+    expectedLocation: path.relative(process.cwd(), config.dirname), // TODO pathToPosix
+  }
+
+  let errorMessage: string
+  if (hasNativeBinaryTarget && hasMissingBinaryTarget) {
+    errorMessage = nativeGeneratedOnDifferentPlatform(errorInput)
+  } else if (hasMissingBinaryTarget) {
+    errorMessage = binaryTargetsWasIncorrectlyPinned(errorInput)
+  } else if (clientHasBeenBundled) {
+    errorMessage = bundlerHasTamperedWithEngineCopy(errorInput)
+  } else {
+    errorMessage = toolingHasTamperedWithEngineCopy(errorInput)
+  }
+
+  throw new PrismaClientInitializationError(errorMessage, config.clientVersion!)
+}
+
+/**
+ * Core logic for the resolution of the engine path. This function will search
+ * for the engine in multiple locations, and return the first one that is found.
+ * @param engineType
+ * @param config
+ * @returns
+ */
+async function findEnginePath(engineType: 'binary' | 'library', config: EngineConfig) {
+  const binaryTarget = await getPlatform()
+  const searchedLocations: string[] = []
+
+  const dirname = eval('__dirname') as string
+  const searchLocations: string[] = [
+    config.dirname, // generation directory
+    path.resolve(dirname, '..'), // generation directory one level up
+    config.generator?.output?.value ?? dirname, // custom generator local path
+    path.resolve(dirname, '../../../.prisma/client'), // dot prisma node_modules ???
+    '/tmp/prisma-engines', // used for netlify
+    config.cwd, // cwdPath, not cwd
+  ]
+
```

Comment:
Just a note: Just noticed that the order is different from before (below is BinaryEngine code)
```ts
    const searchLocations: string[] = [
      eval(`require('path').join(__dirname, '../../../.prisma/client')`), // Dot Prisma Path
      this.generator?.output?.value ?? eval('__dirname'), // Custom Generator Path
      path.join(eval('__dirname'), '..'), // parentDirName
      path.dirname(this.datamodelPath), // Datamodel Dir
      this.cwd, //cwdPath
      '/tmp/prisma-engines',
    ]
```

Reply:
Nice catch, I indeed changed the order. We did a test via ecosystem-tests, and discovered that none of them broke if only `config.dirname` was used (or `path.dirname(this.datamodelPath)`) so it made sense to me to search that one first.